File: modules/google-earth-engine/sepal-ee/sepal/ee/rx/task.py
import logging

import ee
from ee.batch import Task
from rx import empty, of
from rx.operators import distinct_until_changed, finally_action, flat_map, take_while
from sepal.rx import throw
from sepal.task.rx.observables import progress

from .observables import execute, interval

_MONITORING_FREQUENCY = 10


def execute_task(credentials, task):
    def start():

        def action():
            task.start()

        return execute(
            credentials,
            action,
            description='start task ' + str(task),
            retries=3
        )

    def cancel():
        if task.state in [Task.State.COMPLETED, Task.State.CANCEL_REQUESTED, Task.State.CANCELLED, Task.State.FAILED]:
            return
        ee.InitializeThread(credentials)
        try:
            task.cancel()
        except Exception:
            logging.exception('{}: cancelling failed'.format(task))
        return empty()

    def load_status():
        return task.status()

    def extract_state(status):
        state = status['state']
        task.state = state
        if state == 'FAILED':
            return throw(ee.EEException(status.get('error_message')))
        else:
            return of(state)

    def to_progress(state):
        if state == 'PENDING':
            return progress(
                default_message='Submitting export task to Google Earth Engine...',
                message_key='tasks.ee.export.pending'
            )
        elif state == 'READY':
            return progress(
                default_message='Waiting for Google Earth Engine to start export...',
                message_key='tasks.ee.export.ready'
            )
        elif state == 'RUNNING':
            return progress(
                default_message='Google Earth Engine is exporting...',
                message_key='tasks.ee.export.running'
            )
        else:
            return empty()

    def monitor():
        def is_running(state):
            return state in [Task.State.UNSUBMITTED, Task.State.READY, Task.State.RUNNING]

        return interval(credentials, _MONITORING_FREQUENCY).pipe(
            flat_map(
                lambda _: execute(
                    credentials,
                    action=load_status,
                    description='monitor task ' + str(task),
                    retries=3
                )
            ),
            flat_map(extract_state),
            distinct_until_changed(),
            take_while(is_running, inclusive=True),
            flat_map(to_progress)
        )

    return start().pipe(
        flat_map(lambda _: monitor()),
        finally_action(cancel),
    )

File: modules/task/src/ee/task.js
const ee = require('#sepal/ee/ee')
const {interval, of, throwError, catchError, distinctUntilChanged, map, exhaustMap, switchMap, takeWhile, tap} = require('rxjs')
const {finalizeObservable} = require('#sepal/rxjs')
const MONITORING_FREQUENCY = 10000
const {UNSUBMITTED, READY, RUNNING, FAILED} = ee.data.ExportState
const log = require('#sepal/log').getLogger('ee')

const task$ = (taskId, task, description) => {
    const start$ = task =>
        ee.$({
            description: `start task (${description})`,
            operation: (resolve, reject) =>
                ee.data.startProcessing(null, task.config_, (result, error) =>
                    error
                        ? reject(error)
                        : resolve(result.taskId)
                )
        })

    const status$ = (eeTaskId, maxRetries) =>
        ee.$({
            description: `get task status (${description}, ${eeTaskId})`,
            operation: (resolve, reject) =>
                ee.data.getTaskStatus(eeTaskId,
                    (status, error) => error
                        ? reject(error)
                        : resolve(status)
                ),
            maxRetries
        }).pipe(
            map(([status]) => status)
        )

    const cancel$ = (eeTaskId, maxRetries) =>
        ee.$({
            description: `cancel task (${description}, ${eeTaskId})`,
            operation: (resolve, reject) =>
                ee.data.cancelTask(eeTaskId,
                    (_canceled, error) => error
                        ? reject(error)
                        : resolve()
                ),
            maxRetries
        })

    const monitor$ = eeTaskId =>
        interval(MONITORING_FREQUENCY).pipe(
            exhaustMap(() => status$(eeTaskId)),
            switchMap(({state, error_message: error}) =>
                (error || state === FAILED)
                    ? throwError(() => new Error(error))
                    : of(state)
            ),
            distinctUntilChanged(),
            takeWhile(state => isRunning(state)),
            map(toProgress)
        )

    const toProgress = state => {
        switch (state) {
            case 'UNSUBMITTED':
                return {
                    state,
                    defaultMessage: 'Submitting export task to Google Earth Engine',
                    messageKey: 'tasks.ee.export.pending'
                }
            case 'READY':
                return {
                    state,
                    defaultMessage: 'Waiting for Google Earth Engine to start export',
                    messageKey: 'tasks.ee.export.ready'
                }
            case 'RUNNING':
                return {
                    state,
                    defaultMessage: 'Google Earth Engine is exporting',
                    messageKey: 'tasks.ee.export.running'
                }
            default:
                throw Error(`Unknown state (${description}): ${state}`)
        }
    }

    const cleanup$ = eeTaskId => {
        log.debug(() => `EE task cleanup starting (${description}, ${eeTaskId})`)
        return status$(eeTaskId, 0).pipe(
            map(({state}) => isRunning(state)),
            switchMap(running =>
                running
                    ? cancel$(eeTaskId, 3).pipe(
                        map(() => true)
                    )
                    : of(false)
            ),
            tap(wasRunning =>
                log.info(`EE task ${wasRunning ? 'cancelled' : 'completed'} (${description}, ${eeTaskId})`)
            ),
            catchError(error => {
                log.error(`EE task failed to cancel. Trying again, without loading status first (${description}, ${eeTaskId})`, error)
                return cancel$(eeTaskId, 0)
            })
        )
    }

    const isRunning = state => [UNSUBMITTED, READY, RUNNING].includes(state)

    return of(task).pipe(
        switchMap(task => start$(task)),
        switchMap(eeTaskId => monitor$(eeTaskId).pipe(
            finalizeObservable(
                () => cleanup$(eeTaskId),
                taskId,
                `Cleanup EE task ${eeTaskId}`
            )
        ))
    )
}

module.exports = {task$}
